op-node/rollup/derive/batch_queue.go

```diff
@@ -4,205 +4,291 @@ import (
+	bq.l1Blocks = append(bq.l1Blocks, l1BlockStart)
+	bq.progress.Origin = l1BlockStart
+	bq.progress.Closed = false
+	return io.EOF
+}
+
+func (bq *BatchQueue) AddBatch(batch *BatchData) error {
+	if bq.progress.Closed {
+		panic("write batch while closed")
 	}
-	if uint64(len(bq.inputs)) > bq.config.SeqWindowSize {
-		return nil, fmt.Errorf("unexpectedly buffered more L1 inputs than sequencing window: %d", len(bq.inputs))
+	bq.log.Info("queued batch", "origin", bq.progress.Origin, "tx_count", len(batch.Transactions), "timestamp", batch.Timestamp)
+	if len(bq.l1Blocks) == 0 {
+		return fmt.Errorf("cannot add batch with timestamp %d, no origin was prepared", batch.Timestamp)
 	}
-	l1Origin := bq.inputs[0].Origin
-	nextL1Block := bq.inputs[1].Origin
 
-	fetchCtx, cancel := context.WithTimeout(ctx, 20*time.Second)
-	defer cancel()
-	l1Info, _, receipts, err := bq.dl.Fetch(fetchCtx, l1Origin.Hash)
-	if err != nil {
-		bq.log.Error("failed to fetch L1 block info", "l1Origin", l1Origin, "err", err)
-		return nil, nil
+	data := BatchWithL1InclusionBlock{
+		L1InclusionBlock: bq.progress.Origin,
+		Batch:            batch,
 	}
+	batches, ok := bq.batchesByTimestamp[batch.Timestamp]
+	// Filter complete duplicates. This step is not strictly needed as we always append, but it is nice to avoid lots of spam.
+	if ok {
+		for _, b := range batches {
+			if b.Batch.Timestamp == batch.Timestamp && b.Batch.Epoch() == batch.Epoch() {
+				bq.log.Warn("duplicate batch", "epoch", batch.Epoch(), "timestamp", batch.Timestamp, "txs", len(batch.Transactions))
+				return nil
+			}
+		}
+	} else {
+		bq.log.Debug("First seen batch", "epoch", batch.Epoch(), "timestamp", batch.Timestamp, "txs", len(batch.Transactions))
 
-	deposits, errs := DeriveDeposits(receipts, bq.config.DepositContractAddress)
-	for _, err := range errs {
-		bq.log.Error("Failed to derive a deposit", "l1OriginHash", l1Origin.Hash, "err", err)
 	}
-	if len(errs) != 0 {
-		return nil, fmt.Errorf("failed to derive some deposits: %v", errs)
+	// May have duplicate block numbers or individual fields, but have limited complete duplicates
+	bq.batchesByTimestamp[batch.Timestamp] = append(batches, &data)
+	return nil
+}
+
+// validExtension determines if a batch follows the previous attributes
+func validExtension(cfg *rollup.Config, batch *BatchWithL1InclusionBlock, prevTime, prevEpoch uint64) bool {
+	if batch.Batch.Timestamp != prevTime+cfg.BlockTime {
+		return false
 	}
+	if batch.Batch.EpochNum != rollup.Epoch(prevEpoch) && batch.Batch.EpochNum != rollup.Epoch(prevEpoch+1) {
+		return false
+	}
+	// TODO: Also check EpochHash (hard b/c maybe extension)
 
-	minL2Time := uint64(lastL2Timestamp) + bq.config.BlockTime
-	maxL2Time := l1Origin.Time + bq.config.MaxSequencerDrift
-	if minL2Time+bq.config.BlockTime > maxL2Time {
-		maxL2Time = minL2Time + bq.config.BlockTime
+	if uint64(batch.Batch.EpochNum)+cfg.SeqWindowSize < batch.L1InclusionBlock.Number {
```

Comment:
nit: maybe worth a comment that there's no overflow risk due to above check.

Reply:
actually, I'm a little confused, is the attack that the `EpochNum` could be very large when it's submitted?

Reply:
Yes, I think the epoch num is filtered in several other places, but if it affects other batches it might still be an issue. Here we've a check close by that it's one of two non-overflowing values, so it's fine.